### PR TITLE
Add a feature to IngressProcessor to resolve named ports

### LINKED ISSUE #4322: upstream request timeout with ingress v1 when specifying backend.service.port.name
**Describe the bug**

When specifiying ` backend.service.port.name` instead of ` backend.service.port.number`, emissary ingress will not function:

```
HTTP/1.1 504 Gateway Timeout
content-length: 24
content-type: text/plain
date: Fri, 08 Jul 2022 01:45:49 GMT
server: envoy

upstream request timeout
```

**To Reproduce**

1. create a simple pod/service/ingress.

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: test
spec:
  ingressClassName: ambassador
  rules:
  - host: test.example.com
    http:
      paths:
      - backend:
          service:
            name: my-service
            port:
              number: 8080 <<<<<< here
        path: /
        pathType: Prefix
```

This will work fine. Now change the `port.number` and use `port.name` in the service instead. emissary ingress will no longer work for that ingress.

**Expected behavior**

Service backends should work with either `port.number` and `port.name`.

**Versions (please complete the following information):**
 - Ambassador: 3.0.0
 - Kubernetes environment: openshift v4.10
 - Version k8s version 1.23

**Additional context**

I found a TODO here(https://github.com/emissary-ingress/emissary/blob/master/pkg/kates/k8s_resource_types/ingress.go#L20), not sure if it is causing this problem.

As k8s v1.23+ has deprecated and dropped support of ingress v1beta, supporting ingress v1 is essential.

### PR DESCRIPTION
**Pull request by antonu17**
## Problem statement

The Ingress processor ignores the named service port and replaces it with `80`. 

## Related Issues

closes #4322

## Testing

Added a test case to `test_fetcher.py`

## Checklist

- [x] **Does my change need to be backported to a previous release?**
  - would be nice to backport this to v2.4 as well (#4810)

- [ ] **I made sure to update `CHANGELOG.md`.**
- [x] **This is unlikely to impact how Ambassador performs at scale.**
- [x] **My change is adequately tested.**
- [ ] **I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.**
- [ ] **The changes in this PR have been reviewed for security concerns and adherence to security best practices.**

### CODE REVIEW

File: python/ambassador/fetch/ingress.py
Comment (haq204) [inline python/ambassador/fetch/ingress.py:241]: if `_resolve_service_port_number` fails to resolve to a port number then it would return the port name if I'm reading the above method correctly right?

So later on it would default to port 80?

**Review by haq204 — COMMENTED** [review]
thanks for the addressing this!

Just have one question.

File: python/ambassador/fetch/ingress.py
Comment (antonu17) [reply python/ambassador/fetch/ingress.py:241]: that's right, in case the port can't be resolved `_resolve_service_port_number` returns the same `service_port` value. This is possible in a situation if `Service` name specified in the `Ingress` spec doesn't exist, or if the `Service` doesn't have a port with the name provided in the `Ingress`. Both cases are not liable for any ingress controller.

**Comment by haq204** [comment]
@antonu17 I've opened https://github.com/emissary-ingress/emissary/pull/4813 so that CI can run. Our CI is a bit funky in that only maintainers have the proper permissions to run it. 

It appears the only issue is linting. The easiest way to fix that would be to run `make format`

**Comment by antonu17** [comment]
> @antonu17 I've opened #4813 so that CI can run. Our CI is a bit funky in that only maintainers have the proper permissions to run it.
> 
> It appears the only issue is linting. The easiest way to fix that would be to run `make format`

Got it, pushed `make format` changes

**Comment by haq204** [comment]
> > @antonu17 I've opened #4813 so that CI can run. Our CI is a bit funky in that only maintainers have the proper permissions to run it.
> > It appears the only issue is linting. The easiest way to fix that would be to run `make format`
> 
> Got it, pushed `make format` changes

Looks like mypy is still not happy...
```
python/tests/unit/test_fetch.py:331: error: Argument 1 to "len" has incompatible type "Optional[Dict[str, ACResource]]"; expected "Sized"
python/tests/unit/test_fetch.py:333: error: Item "None" of "Optional[Dict[str, ACResource]]" has no attribute "values"
```

**Comment by antonu17** [comment]
@haq204 Thanks for re-trying CI, i see it still fails with:

```
python/tests/unit/test_fetch.py:331: error: Argument 1 to "len" has incompatible type "Optional[Dict[str, ACResource]]"; expected "Sized"
python/tests/unit/test_fetch.py:333: error: Item "None" of "Optional[Dict[str, ACResource]]" has no attribute "values"
Found 2 errors in 1 file (checked 189 source files)
```

I think I understand why. Pushed another commit that fixes it.

**Comment by haq204** [comment]
> @haq204 Thanks for re-trying CI, i see it still fails with:
> 
> ```
> python/tests/unit/test_fetch.py:331: error: Argument 1 to "len" has incompatible type "Optional[Dict[str, ACResource]]"; expected "Sized"
> python/tests/unit/test_fetch.py:333: error: Item "None" of "Optional[Dict[str, ACResource]]" has no attribute "values"
> Found 2 errors in 1 file (checked 189 source files)
> ```
> 
> I think I understand why. Pushed another commit that fixes it.

@antonu17 thanks! I updated the other PR for changelog updates and rebased it onto master. CI is green there so once it gets approvals from the other maintainers it should be good to merge.

> @haq204 Thanks for re-trying CI, i see it still fails with:
> 
> ```
> python/tests/unit/test_fetch.py:331: error: Argument 1 to "len" has incompatible type "Optional[Dict[str, ACResource]]"; expected "Sized"
> python/tests/unit/test_fetch.py:333: error: Item "None" of "Optional[Dict[str, ACResource]]" has no attribute "values"
> Found 2 errors in 1 file (checked 189 source files)
> ```
> 
> I think I understand why. Pushed another commit that fixes it.

**Comment by antonu17** [comment]
> @antonu17 thanks! I updated the other PR for changelog updates and rebased it onto master. CI is green there so once it gets approvals from the other maintainers it should be good to merge.

@haq204 thanks for your review. should I close this PR?

File: python/ambassador/fetch/ingress.py
Comment (LanceEa) [inline python/ambassador/fetch/ingress.py:108]: discovered_services is a dict, can't we just use `self.service_dep.discovered_services.get(f"{service_name}.{namespace}", None)`? 

If the cluster has a larger number of services we potentially would have to do a lot of looping here.

**Review by LanceEa — CHANGES_REQUESTED** [review]
@antonu17 @haq204 - Thanks for this and this will be super useful for folks using Ingress resources. 

I left a few notes on the code and test. I need familiarize myself a little more with this part of the code base so some of my comments/questions might be obvious once I dive deeper into it.

File: python/ambassador/fetch/dependency.py
Comment (antonu17) [reply python/ambassador/fetch/dependency.py:39]: I saw this code for the first time in my live this week xD, sorry in advance if I misunderstood something.

So `ambassador_service` seems to be ambassador's loadbalancer service which is used for referencing in Ingress status info.
And `discovered_services` dictionary used to be a field of the `InternalServiceProcessor`, but I moved it to `ServiceDependency` class to make it available for IngressProcessor for port lookups.

File: python/ambassador/fetch/ingress.py
Comment (antonu17) [reply python/ambassador/fetch/ingress.py:108]: My bad 🤦  will update soon

File: python/tests/unit/test_fetch.py
Comment (antonu17) [reply python/tests/unit/test_fetch.py:282]: Alright, I'll add more test cases to address these additional cases

**Comment by antonu17** [comment]
@LanceEa, @haq204 I addressed all PR comments and updated the code and tests. have a look once more please.

File: python/ambassador/fetch/ingress.py
Comment (haq204) [inline python/ambassador/fetch/ingress.py:121]: since line 119 also could also return the service port, this debug log wouldn't always get outputted if `service_port` is returned

**Review by haq204 — COMMENTED** [review]
one nit but otherwise lgtm once @LanceEa takes a look

File: python/ambassador/fetch/ingress.py
Comment (antonu17) [reply python/ambassador/fetch/ingress.py:121]: Correct. If the port is found we don't need to display "not found" message

File: python/tests/unit/test_fetch.py
Comment (LanceEa) [inline python/tests/unit/test_fetch.py:374]: maybe I'm missing something but aren't we trying to resolve the names to port numbers here? 

I would expect this one to be rejected and dropped or fallback to something, same with the metrics above too.

Comment (antonu17) [reply python/tests/unit/test_fetch.py:374]: Hello, @LanceEa! 
The fallback behavior is to retain the named port in the `aconf` mapping structure (this is how the ingress fetcher works at the moment). The bad port will be replaced to 80/443 later here: https://github.com/emissary-ingress/emissary/blob/aee9d485917365a2efa53ce90f121483da5c95a0/python/ambassador/ir/ircluster.py#L212-L215

Comment (antonu17) [reply python/tests/unit/test_fetch.py:374]: The point of these additional ingress fetcher tests is to demonstrate that port number resolution works whenever it is possible: referenced service and ports exist. In case of ingress' paths reference non-existing services or ports fetcher behaves as before.
Unfortunately, there were no tests for ingress fetcher before, so there's nothing to compare these tests with.
One thing I can do to better emphasize that the port number resolution is not guaranteed is to add the word `try` to the name of the method.

Comment (LanceEa) [reply python/tests/unit/test_fetch.py:374]: ahh got it. Like I mentioned, I wasn't as familiar with this part of the codebase 😄. My initial gut instinct was to assume that the resolution would have happened at this stage rather than delaying until later stages. So, LGTM

**Review by LanceEa — APPROVED** [review]

**Comment by haq204** [comment]
@antonu17 can you rebase onto `main`? Once that is done we can run CI and get it merged

**Comment by haq204** [comment]
@antonu17 CI in https://github.com/emissary-ingress/emissary/pull/4813 is green and has been approved and merged.

I'm going to close this one out. 

Thanks again for your contribution!

**Pull request closed without merging**
